cmd/chisel/db.go

```diff
@@ -0,0 +1,250 @@
+
+	generatedPaths := []string{}
+	for path, dbw := range dbWriters {
+		// Add packages to the DB.
+		for _, info := range opts.PackageInfo {
+			err := dbw.Add(&Package{
+				Kind:    "package",
+				Name:    info.Name,
+				Version: info.Version,
+				Digest:  info.Hash,
+				Arch:    info.Arch,
+			})
+			if err != nil {
+				return nil, err
+			}
+		}
+		// Add slices to the DB.
+		for _, s := range opts.Slices {
+			err := dbw.Add(&Slice{
+				Kind: "slice",
+				Name: s.String(),
+			})
+			if err != nil {
+				return nil, err
+			}
+		}
+		// Add paths and contents to the DB.
+		for _, entry := range opts.Report.Entries {
+			mode := fmt.Sprintf("0%o", entry.Mode&fs.ModePerm)
+			sliceNames := []string{}
+			for s := range entry.Slices {
+				name := s.String()
+				// Add contents to the DB.
+				err := dbw.Add(&Content{
+					Kind:  "content",
+					Slice: name,
+					Path:  entry.Path,
+				})
+				if err != nil {
+					return nil, err
+				}
+				sliceNames = append(sliceNames, name)
+			}
+			err := dbw.Add(&Path{
+				Kind:   "path",
+				Path:   entry.Path,
+				Mode:   mode,
+				Slices: sliceNames,
+				Hash:   entry.Hash,
+				Size:   uint64(entry.Size),
+				Link:   entry.Link,
+			})
+			if err != nil {
+				return nil, err
+			}
+		}
+		// Add the DB path and content entries.
+		for _, pathEntry := range dbPathEntries {
+			err := dbw.Add(pathEntry)
+			if err != nil {
+				return nil, err
+			}
+			for _, s := range pathEntry.Slices {
+				err := dbw.Add(&Content{
+					Kind:  "content",
+					Slice: s,
+					Path:  pathEntry.Path,
+				})
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+
+		dbPath := dbPaths[path]
+		logf("Generating Chisel DB at %s...", dbPath)
+		err := WriteDB(dbw, dbPath, dbMode)
+		if err != nil {
+			return nil, err
+		}
+		relPath := filepath.Clean("/" + strings.TrimPrefix(dbPath, opts.RootDir))
+		generatedPaths = append(generatedPaths, relPath)
+	}
+
+	return generatedPaths, nil
+}
+
+type Package struct {
+	Kind    string `json:"kind"`
+	Name    string `json:"name"`
+	Version string `json:"version"`
+	Digest  string `json:"sha256"`
+	Arch    string `json:"arch"`
+}
+
+type Slice struct {
+	Kind string `json:"kind"`
+	Name string `json:"name"`
+}
+
+type Path struct {
+	Kind      string   `json:"kind"`
+	Path      string   `json:"path"`
+	Mode      string   `json:"mode"`
+	Slices    []string `json:"slices"`
+	Hash      string   `json:"sha256,omitempty"`
+	FinalHash string   `json:"final_sha256,omitempty"`
+	Size      uint64   `json:"size,omitempty"`
+	Link      string   `json:"link,omitempty"`
+}
+
+type Content struct {
+	Kind  string `json:"kind"`
+	Slice string `json:"slice"`
+	Path  string `json:"path"`
+}
+
+// WriteDB writes all added entries to the Chisel DB and generates the actual
+// file. It returns the path of the generated Chisel DB file. The file
+// chisel.db is a zstd compressed file.
+func WriteDB(writer *jsonwall.DBWriter, path string, mode fs.FileMode) (err error) {
```

Comment:
If the mode is going to be hardcoded to 644, let's remove the argument. We can always add it back later. See my comment below about how to refactor the Writer. If we are doing that, I think the function structure could be further simplified by opening the files inline, then creating the multiwriter and then the zstd. Lastly, we are already handling the jsonwall.DBWriter so we can call WriteTo directly.

You have more context about these changes than I do so if there is another layout that feels more natural, keep it like that.

Reply:
Removed the `mode` argument. But kept the `WriteDB` function though. I feel that keeping the actual I/O operations separately will be beneficial -- no functions are huge, we can easily change the I/O ops later etc.

```go
func WriteDB(writer *jsonwall.DBWriter, paths []string) (err error)
```